README.md
# advent-of-code
Clojure Adventures with Advent of Code 2019


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-12-06
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-12-06
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/advent-of-code/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/advent-of-code/compare/0.1.0...0.1.1


test/adventofcode/day02_test.clj
(ns adventofcode.day02-test
  (:require [clojure.test :refer :all]
            [adventofcode.day02 :refer :all]))

(deftest test-get-opcode
  (testing "get-opcode"
    (testing "returns opcode stored at instruction pointer"
      (is (= 1 (get-opcode [1,2,3,4] 0)))
      (is (= 2 (get-opcode [1,2,3,4] 1)))
      (is (= 3 (get-opcode [1,2,3,4] 2)))
      (is (= 4 (get-opcode [1,2,3,4] 3))))))

(deftest test-calculate-next-pointer
  (testing "next-pointer increments pointer by 4"
    (is (= 4 (calculate-next-pointer 0)))
    (is (= 8 (calculate-next-pointer 4)))))

(deftest test-output-pos
  (testing "output-pos"
    (testing "returns the forth value of an intcode"
      (is (= 4 (output-pos [1,2,3,4] 0)))
      (is (= 2 (output-pos [3,2,1,2] 0)))
      (is (= 6 (output-pos [4,6,2,6] 0))))

    ; The output position is always the value at offset 3 starting at the pointer
    (testing "considers the intcode pointer"
      (is (= 1 (output-pos [1,2,3,4,1,2,3] 1)))
      (is (= 4 (output-pos [1,2,3,4,1,4,0] 2)))
      (is (= 9 (output-pos [1,2,3,4,1,8,9] 3))))))

(deftest test-get-input-value-at
  (testing "get-input-value-at returns the value of the field denoted by the index"
    (is (= 10 (get-input-value-at [1,4,5,4,10,20,30] 0 1)))
    (is (= 20 (get-input-value-at [1,4,5,4,10,20,30] 0 2)))
    (is (= 2  (get-input-value-at [1,4,5,4,2,4,5,1] 4 1)))
    (is (= 4  (get-input-value-at [1,4,5,4,2,4,5,1] 4 2)))))

(deftest test-run-program
  (testing "Running the program with pre-defined values from the instructions."
    (is (= [2 0 0 0 99] (run-program [1,0,0,0,99])))
    (is (= [2 3 0 6 99] (run-program [2,3,0,3,99])))
    (is (= [2 4 4 5 99 9801] (run-program [2,4,4,5,99,0])))
    (is (= [30 1 1 4 2 5 6 0 99] (run-program [1,1,1,4,99,5,6,0,99])))))

(deftest test-restore-1202-program-alarm-state
  (testing "Fixes 1202 program seed state"
    (is (= [1,12,2,0,1,2,3,4] (restore-1202-program-alarm-state [1,2,3,0,1,2,3,4])))
    (is (= [99,12,2,1] (restore-1202-program-alarm-state [99,1,1,1])))))

(deftest test-prepare-program-seeds
  (testing "prepare-program-seeds injects given seed values into program"
    (is (= [1,2,3] (prepare-program-seeds [1,0,0] 2 3)))
    (is (= [1,2,3,0,0] (prepare-program-seeds [1,0,0,0,0] 2 3)))
    (is (= [1,99,0,0,0,99] (prepare-program-seeds [1,0,0,0,0,99] 99 0)))))

(deftest test-create-program-seeds
  (testing "Creating program seeds produces values"
    (is (= '((0 0) (1 0) (2 0) (0 1) (1 1) (2 1) (0 2) (1 2) (2 2)) (create-program-seeds 3 3)))
    (is (= '((0 0) (1 0)) (create-program-seeds 2 1)))
    ))

(deftest test-find-program-seeds
  (testing "Checks that find-program-seeds produces the originally used 1202 code."
    (is (= [12, 2] (find-program-seeds (read-program) 3409710)))))

test/adventofcode/day03_test.clj
(ns adventofcode.day03-test
  (:require [clojure.test :refer :all]
            [adventofcode.day03 :refer :all]))


(deftest test-scalar-distance
  (testing "Scalar distance of values"
    (is (= 0 (scalar-distance 0 0)))
    (is (= 0 (scalar-distance 1 1)))
    (is (= 0 (scalar-distance 10 10)))
    (is (= 10 (scalar-distance 10 0)))
    (is (= 10 (scalar-distance 0 10)))
    (is (= 5 (scalar-distance 2 7)))
    (is (= 5 (scalar-distance 7 2)))))

(deftest test-horizontal-distance
  (testing "Calculate horizontal distance"
    (is (= 0  (horizontal-distance [0  1] [0 0])))
    (is (= 0 (horizontal-distance  [0 10] [0 0])))
    (is (= 2 (horizontal-distance  [4  8] [6 0])))
    (is (= 2 (horizontal-distance  [6  0] [4 8])))
    (is (= 0  (horizontal-distance [6  8] [6 8])))))

(deftest test-vertical-distance
  (testing "Calculate manhattan distance"
    (is (= 1  (vertical-distance [0  1] [0 0])))
    (is (= 10 (vertical-distance [0 10] [0 0])))
    (is (= 8  (vertical-distance [4  8] [6 0])))
    (is (= 0  (vertical-distance [6  8] [6 8])))))

(deftest test-manhattan-distance
  (testing "Calculate manhattan distance"
    (is (= 1  (manhattan-distance [0  1] [0 0])))
    (is (= 10 (manhattan-distance [0 10] [0 0])))
    (is (= 10 (manhattan-distance [4  8] [6 0])))
    (is (= 0  (manhattan-distance [6 8] [6 8])))))

(deftest test-get-steps
  (testing "Extracts steps"
    (is (= 10 (get-steps "R10")))
    (is (= 0 (get-steps "R0")))
    (is (= 3 (get-steps "R3")))))

(deftest test-produce-steps-from
  (testing "Step coordinates right"
    (is (= [[0 0] [1 0] [2 0]] (produce-steps-from [0 0] "R2")))
    (is (= [[0 0] [1 0] [2 0] [3 0]] (produce-steps-from [0 0] "R3")))
    (is (= [[10 5] [11 5] [12 5] [13 5]] (produce-steps-from [10 5] "R3"))))

  (testing "Step coordinates right"
    (is (= [[0 0] [-1 0] [-2 0]] (produce-steps-from [0 0] "L2")))
    (is (= [[0 0] [-1 0] [-2 0] [-3 0]] (produce-steps-from [0 0] "L3")))
    (is (= [[10 5] [9 5] [8 5] [7 5]] (produce-steps-from [10 5] "L3"))))

  (testing "Step coordinates down"
    (is (= [[0 0] [0 1] [0 2]] (produce-steps-from [0 0] "D2")))
    (is (= [[0 0] [0 1] [0 2] [0 3]] (produce-steps-from [0 0] "D3")))
    (is (= [[10 5] [10 6] [10 7]] (produce-steps-from [10 5] "D2"))))

  (testing "Step coordinates up"
    (is (= [[0 0] [0 -1] [0 -2]] (produce-steps-from [0 0] "U2")))
    (is (= [[0 0] [0 -1] [0 -2] [0 -3]] (produce-steps-from [0 0] "U3")))
    (is (= [[10 5] [10 4] [10 3]] (produce-steps-from [10 5] "U2")))))

(deftest test-produce-steps-from
  (testing "Create steps from"
    (is (= [[0 0] [1 0] [2 0]] (produce-steps-from [0 0] "R2")))
    (is (= [[0 0] [1 0] [2 0] [3 0]] (produce-steps-from [0 0] "R3")))))

(deftest test-create-wire
  (testing "Turns wire commands into wire"
    (is (= [[0 0] [1 0] [2 0]] (create-wire ["R2"])))
    (is (= [[0 0] [1 0] [2 0] [2 -1] [2 -2]] (create-wire ["R2", "U2"]))))

  (testing "Walking in circles"
    (is (= [0 0] (last (create-wire ["R2", "U2", "L2", "D2"]))))))

(deftest test-find-wire-intersections
  (testing "Wires without intersections"
    (is (= #{} (find-intersections [[0 0] [0 1]] [[1 0] [1 1]])))
    (is (= #{[0 1]} (find-intersections [[0 0] [0 1]] [[1 0] [1 1] [0 1]])))))

(deftest test-get-wire-length-to
  (testing "Returns wire length to a given coordinate"
    (is (= 0 (get-wire-length-to [[0 0]] [0 0])))
    (is (= 3 (get-wire-length-to [[0 0] [1 0] [2 0] [2 1]] [2 1]))))

  (testing "Returns nil if cordinate not in wire"
    (is (nil? (get-wire-length-to [[0 0] [1 0] [2 0] [2 1]] [2 2])))))

test/adventofcode/day04_test.clj
(ns adventofcode.day04-test
  (:require [clojure.test :refer :all]
            [adventofcode.day04 :refer :all]))

(deftest test-has-doubles-or-larger
  (testing "Does have adjacent digits that are the same"
    (is (true? (has-doubles-or-larger? 11)))
    (is (true? (has-doubles-or-larger? 122)))
    (is (true? (has-doubles-or-larger? 121445)))
    (is (true? (has-doubles-or-larger? 1111111))))

  (testing "Does have adjacent digits of three which are the same"
    (is (true? (has-doubles-or-larger? 111)))
    (is (true? (has-doubles-or-larger? 1222)))
    (is (true? (has-doubles-or-larger? 1214445)))
    (is (true? (has-doubles-or-larger? 1111111))))

  (testing "Does not have adjacent digits that are the same"
    (is (not (true? (has-doubles-or-larger? 10))))
    (is (not (true? (has-doubles-or-larger? 10101010))))
    (is (not (true? (has-doubles-or-larger? 123456))))))

(deftest test-monotonical-non-decreasing-digits
  (testing "All digits are monotonical non decreasing"
    (is (true? (monotononical-non-decreasing-digits? 11)))
    (is (true? (monotononical-non-decreasing-digits? 111)))
    (is (true? (monotononical-non-decreasing-digits? 12345)))
    (is (true? (monotononical-non-decreasing-digits? 1122333444455555))))

  (testing "Some digits decrease"
    (is (not (true? (monotononical-non-decreasing-digits? 10))))
    (is (not (true? (monotononical-non-decreasing-digits? 121))))
    (is (not (true? (monotononical-non-decreasing-digits? 1234345))))))

(deftest test-split-number-into-digits
  (testing "Splits number into collections of digits"
    (is (= [1] (split-number-into-digits 1)))
    (is (= [1 1] (split-number-into-digits 11)))
    (is (= [1 2 3 4 5] (split-number-into-digits 12345)))))

(deftest test-has-doubles
  (testing "Number has doubles"
    (is (true? (has-doubles? 11)))
    (is (true? (has-doubles? 1221)))
    (is (true? (has-doubles? 122211))))

  (testing "Triples don't count as doubles"
    (is (not (true? (has-doubles? 111))))
    (is (not (true? (has-doubles? 111222))))
    (is (not (true? (has-doubles? 12333))))))

src/adventofcode/day03.clj
; day03
(ns adventofcode.day03
  (:require [clojure.string :as str])
  (:require [clojure.edn :as edn]))

; Generic
(defn x "x component of two dimensional carthesian vector" [point] (first point))
(defn y "y component of two dimensional carthesian vector" [point] (second point))

; Distance calculations
(defn scalar-distance
  "Calculates the distance between the two given values."
  [scalar1 scalar2] (Math/abs (- scalar1 scalar2)))

(defn component-distance [point1 point2 component]
  (scalar-distance (component point1) (component point2)))

(defn horizontal-distance
  "Calculates the horizontal distance between the two points."
  [point1 point2] (component-distance point1 point2 x))

(defn vertical-distance
  "Calculates the vertical distance between the two points"
  [point1 point2] (component-distance point1 point2 y))

(defn manhattan-distance
  "Manhattan distance of the two given points."
  [p1 p2]
  (+ (horizontal-distance p1 p2) (vertical-distance p1 p2)))


; Wire creation
(defn get-steps [command]
  (edn/read-string (subs command 1)))

(defn produce-steps-from
  [start-coordinate command]
  (let [[x y] start-coordinate
        direction (subs command 0 1)
        steps (get-steps command)]
    (case direction
      "U" (partition 2
                     (interleave
                      (repeat (+ steps 1) x)
                      (range y (- y steps 1) -1)))
      "D" (partition 2
                     (interleave
                      (repeat (+ steps 1) x)
                      (range y (+ y steps 1) 1)))
      "R" (partition 2
                     (interleave
                      (range x (+ x steps 1) 1)
                      (repeat (+ steps 1) y)))
      "L" (partition 2
                     (interleave
                      (range x (- x steps 1) -1)
                      (repeat (+ steps 1) y))))))

(defn create-wire
  ([command-list] (create-wire [[0 0]] command-list))
  ([current-wire command-list]
   (if (empty? command-list)
     current-wire
     (let [command (first command-list)
           end-point (last current-wire)
           new-wire (concat current-wire
                            (rest (produce-steps-from end-point command)))]
       (create-wire new-wire (rest command-list))))))

; Wire testing
(defn find-intersections
  [wire1 wire2]
  (let [wire-set1 (set wire1)
        wire-set2 (set wire2)]
    (clojure.set/intersection wire-set1 wire-set2)))

; Input parsing
(defn read-input-file
  ([] (read-input-file "day3.txt"))
  ([filename] (slurp filename)))

(defn parse-wires
  "Takes the raw input file content and returns the actual wires in coordinates"
  [file-content]
  (->> (str/split-lines file-content)
       (map #(str/split % #","))
       (map #(create-wire %))))

; Program logic

; day03, part 1

(defn day3-part1 []
  (let [wires (parse-wires (read-input-file))]
    (second
     (sort
      (map (partial manhattan-distance [0 0])
           (find-intersections (first wires) (second wires)))))))

; day03, part 2

(defn get-wire-length-to
  "Returns the length of the wire until coordinate or nil if the coordinate
   is not part of the wire."
  [wire coordinate]
  (let [index (.indexOf wire coordinate)]
    (if (> index -1)
      index
      nil)))

(defn day3-part2 []
  (let [wires (parse-wire-command (read-input-file))
        len-calculator-1 (partial get-wire-length-to (first wires))
        len-calculator-2 (partial get-wire-length-to (second wires))]
    (second
     (sort
      (map #(+ (len-calculator-1 %) (len-calculator-2 %))
           (find-intersections (first wires) (second wires)))))))

(day3-part2)

src/adventofcode/day04.clj
; day04
(ns adventofcode.day04)

(def START 152085)
(def END   670283)

(defn has-groups-matching
  "Returns true if coll contains at least one group having the given size."
  ([predicate coll]
   (some predicate coll)))

(defn split-number-into-digits
  "Converts the given number into a sequence of its digits."
  [number] (map #(Character/digit % 10) (str number)))

(defn monotononical-non-decreasing-digits? 
  [number]
  "Checks any two adjacent digits are either in increasing order or the same"
  (->> (split-number-into-digits number)
       (partition 2 1)
       (every? #(apply <= %))))
  
(defn has-doubles-or-larger?
  [number]
  (->> (split-number-into-digits number)
       (partition-by identity)
       (has-groups-matching #(<= 2 (count %)))))

(defn has-doubles?
  "Tests if number has double digits in it. Tripple or larger groups don't count."
  [number]
  (->> (split-number-into-digits number)
       (partition-by identity)
       (has-groups-matching #(= 2 (count %)))))

(defn day04-part1 []
  (count
   (filter (every-pred
            has-doubles-or-larger?
            monotononical-non-decreasing-digits?)
           (range START (inc END)))))

(defn day04-part2 []
  (count
   (filter (every-pred
            has-doubles?
            monotononical-non-decreasing-digits?)
           (range START (inc END)))))

(day04-part2)

src/adventofcode/day05.clj
(ns adventofcode.day05)

(require '[clojure.string :as str])

(def OP_ADD 1)
(def OP_MULTIPLY 2)
(def OP_INPUT 3)
(def OP_OUTPUT 4)
(def OP_EXIT 99)
(def PROGRAM_STEP 4)

(def PARAM_MODE_IMMEDIATE 1)
(def PARAM_MODE_ADDRESS PARAM_MODE_IMMEDIATE)
(def PARAM_MODE_POSITION 0)


(def INPUT
  "Input value, which we provide to the only input operation in the TEST program"
  [1])

(defn split-number-into-digits
  "Converts the given number into a sequence of its digits."
  [number] (map #(Character/digit % 10) (str number)))

(defn read-program
  "Returns the parsed program"
  []
  (let [program (slurp "day5.txt")]
    (->> (str/split program #",")
         (mapv #(Integer/parseInt %1)))))

(defn second-last
  "Returns the second but last item in coll in linear time."
  [coll] (last (butlast coll)))

(defn get-param
  [param-pos program pointer param-mode]
  (condp some [param-mode]
    #{PARAM_MODE_ADDRESS PARAM_MODE_IMMEDIATE} (nth program (+ pointer param-pos 1))
    #{PARAM_MODE_POSITION} (let [address (nth program (+ pointer param-pos 1))]
                             (nth program address))))
(defn parse-parameters
  [program pointer parameter-modes operation]
  (condp some [operation]
    #{OP_ADD OP_MULTIPLY} [(get-param 0 program pointer (last parameter-modes))
                           (get-param 1 program pointer (second-last parameter-modes))
                           (get-param 2 program pointer PARAM_MODE_ADDRESS)]

; WHAT SHOULD WE DO HERE? Is this line correct?
    #{OP_INPUT OP_OUTPUT} [(get-param 0 program pointer (last parameter-modes))])

(defn parse-operation
  "Reads the opcode and resolved parameters from the program at the given pointer."
  [program pointer]
  (let [opcode-vector (split-number-into-digits (nth program pointer))
        operation (clojure.string/join (take-last 2 opcode-vector))
        parameter-modes (take (- (count opcode-vector) 2) opcode-vector)
        parameters (parse-parameters program
                                     pointer
                                     parameter-modes
                                     operation)]
    [operation parameters]))


(defn get-input-value-at
  "Given a program and a program pointer, this methods resolves 
   the field which is addressed by the input value at index."
  [program pointer index]
  (nth program (nth program (+ pointer index))))

(defn output-pos
  [program pointer]
  (nth program (+ pointer 3)))

(defn calculate-next-pointer
  "Returns the next incode pointer which should be executed after the
   current pointer."
  [pointer opcode]
  (let [step-size (condp some [opcode]
                    #{OP_ADD OP_MULTIPLY} 4
                    #{OP_INPUT OP_OUTPUT} 2)]
    (+ pointer step-size)))

(defn execute-op-at
  "Takes a program, a frame pointer and an operation (+ or *) which
   should be performed in the current frame. Returns the resulting 
   program."
  [program pointer op]
  (assoc program
         (output-pos program pointer)
         (op (get-input-value-at program pointer 1)
             (get-input-value-at program pointer 2))))

(defn add [program pointer] (execute-op-at program pointer +))
(defn multiply [program pointer] (execute-op-at program pointer *))

(declare execute-op-input
         execute-op-output)

(defn run-program
  "Recursively execute the program, returning the final program 
   as soon as the exit code 99 is reached."
  ([program input] (run-program program 0 input []))
  ([program pointer input output]
   (let [[opcode parameters] (parse-operation program pointer)
         next-pointer (calculate-next-pointer pointer opcode)]
     (condp = opcode
       OP_ADD  (run-program (add program pointer) next-pointer)
       OP_MULTIPLY (run-program (multiply program pointer) next-pointer)
       OP_INPUT (execute-op-input program pointer input output)
       OP_OUTPUT (execute-op-output program pointer input output)
       OP_EXIT output))))

(defn do-input [program pointer input-value])
(defn do-output [program pointer output])

(defn execute-op-input
  [program pointer input output]
  ; (run-program (do-input program pointer (first input)) next-pointer (rest input) output))
  ())

(defn execute-op-output
  [program pointer input output]
  ; (run-program program next-pointer input (do-output program pointer output)))
  ())

(defn day5-part1
  "Takes input and program, and returns the output values of the TEST program."
  []
  (run-program (read-program) INPUT))

src/adventofcode/day02.clj
; Day 2, part 1
(ns adventofcode.day02)

(require '[clojure.string :as str])

(def OP_ADD 1)
(def OP_MULTIPLY 2)
(def OP_EXIT 99)
(def PROGRAM_STEP 4)

(defn read-program []
  (let [program (slurp "day2.txt")]
    (->> (str/split program #",")
         (mapv #(Integer/parseInt %1)))))

(defn get-opcode
  "Reads the opcode from the program at the given pointer."
  [program pointer]
  (nth program pointer))

(defn get-input-value-at
  "Given a program and a program pointer, this methods resolves 
   the field which is addressed by the input value at index."
  [program pointer index]
  (nth program (nth program (+ pointer index))))

(defn output-pos
  [program pointer]
  (nth program (+ pointer 3)))

(defn calculate-next-pointer
  "Returns the next incode pointer which should be executed after the
   current pointer."
  [pointer]
  (+ pointer PROGRAM_STEP))

(defn execute-op-at
  "Takes a program, a frame pointer and an operation (+ or *) which
   should be performed in the current frame. Returns the resulting 
   program."
  [program pointer op]
  (assoc program
         (output-pos program pointer)
         (op (get-input-value-at program pointer 1)
             (get-input-value-at program pointer 2))))

(defn add [program pointer] (execute-op-at program pointer +))
(defn multiply [program pointer] (execute-op-at program pointer *))

(defn run-program
  "Recursively execute the program, returning the final program 
   as soon as the exit code 99 is reached."
  ([program] (run-program program 0))
  ([program pointer]
   (let [opcode (get-opcode program pointer)
         next-pointer (calculate-next-pointer pointer)]
     (case opcode
       1  (run-program (add program pointer) next-pointer)
       2  (run-program (multiply program pointer) next-pointer)
       99 program))))

(defn prepare-program-seeds
  [program seed1 seed2]
  (-> program
      (assoc 1 seed1)
      (assoc 2 seed2)))

(defn get-program-result [program] (first program))

; day 2; example 1
(defn restore-1202-program-alarm-state
  [program]
  (prepare-program-seeds program 12 2))

(defn load-and-run-program-with-seeds
  [seed1 seed2]
  (get-program-result
   (run-program
    (prepare-program-seeds
     (read-program) seed1 seed2))))

(defn day2-part1 []
  (get-program-result
   (run-program
    (restore-1202-program-alarm-state
     (read-program)))))

;(day2-part1)

; day 2; example 2

(defn create-program-seeds
  "Creates a list of seed value pairs"
  [x y]
  (partition 2
             (interleave
              (take (* x y) (cycle (range 0 x)))
              (mapcat #(repeat x %) (range 0 y)))))

(defn seed-yields-result-factory
  [program result] 
  #(let [[seed1 seed2] %]
      (= result (get-program-result (run-program
                                     (prepare-program-seeds program seed1 seed2))))))

; Better approach than this: Create a lazy sequence of seed pairs, 
; iterate over the list, and stop with the result.
(defn find-program-seeds
  "Finds and returns necessary seeds for 'program' that procudes 'output'
   by trying different program seeds values. Seed values is a vector
   containing two values. Returns a vector with two values."
  [program desired-output]
  (let [seed-values (create-program-seeds 100 100)]
    (first (filter (seed-yields-result-factory program desired-output) seed-values))))

(find-program-seeds (read-program) 193409710690720)